ocrd_network/ocrd_network/processing_server.py

```diff
@@ -554,63 +587,59 @@ async def get_processor_job(self, processor_name: str, job_id: str) -> PYJobOutp
-        job_page_ids = expand_page_ids(job_db.page_id)
+        self.log.debug(f"Result job_id: {result_job_id}, state: {result_job_state}")
 
         # Read DB workspace entry
-        workspace_db = await db_get_workspace(
-            workspace_id=workspace_id,
-            workspace_mets_path=path_to_mets
-        )
-        if not workspace_db:
+        db_workspace = await db_get_workspace(workspace_id=workspace_id, workspace_mets_path=path_to_mets)
+        if not db_workspace:
             self.log.exception(f"Workspace with id: {workspace_id} or path: {path_to_mets} not found in DB")
 
-        locked_ws_pages = workspace_db.pages_locked
-        # Update locked pages by unlocking the pages in the request
-        self.unlock_pages(
-            locked_ws_pages=locked_ws_pages,
-            output_file_grps=job_output_file_grps,
-            page_ids=job_page_ids
-        )
-
-        # Update the locked pages dictionary in the database
-        await db_update_workspace(
-            workspace_id=workspace_id,
-            workspace_mets_path=path_to_mets,
-            pages_locked=locked_ws_pages
-        )
-
-        # Take the next request from the cache (if any available)
         workspace_key = workspace_id if workspace_id else path_to_mets
-
+        # Take the next request from the cache (if any available)
         if workspace_key not in self.processing_requests_cache:
             self.log.debug(f"No internal queue available for workspace with key: {workspace_key}")
             return
 
+        if result_job_state == StateEnum.failed:
+            if len(self.processing_requests_cache[workspace_key]):
+                self.log.debug(f"Cancelling jobs dependent to job_id: {result_job_id}")
+                await self.cancel_dependent_jobs(
+                    job_id=result_job_id,
+                    internal_queue=self.processing_requests_cache[workspace_key]
+                )
+
+        if result_job_state != StateEnum.success:
+            # TODO: Handle other potential error cases
+            pass
+
+        db_result_job = await db_get_processing_job(result_job_id)
+        if not db_result_job:
+            self.log.exception(f"Processing job with id: {result_job_id} not found in DB")
+
+        # Unlock the output file group pages for the result processing request
+        await self.unlock_pages(
+            db_workspace=db_workspace,
+            output_file_grps=db_result_job.output_file_grps,
+            page_ids=expand_page_ids(db_result_job.page_id)
+        )
+
         if not len(self.processing_requests_cache[workspace_key]):
             # The queue is empty - delete it
             try:
                 del self.processing_requests_cache[workspace_key]
             except KeyError:
                 self.log.warning(f"Trying to delete non-existing internal queue with key: {workspace_key}")
+
+            # Shut down the Mets Server for the workspace_key
```

Comment:
See main discussion. IIUC we have discussed three solutions

1. Highjack the callback mechanism to be notified when all jobs are finished
2. Have an inactivity timeout in the METS server to automatically shut down when no more files are added
3. Poll the number of jobs completed/failed and shutdown once it reaches 0

Solution 1) is difficult to implement, especially when doing page-wise processing because the pages do not depend on each other (?)

Solution 2) would make sense in any case because there can always be situations where it is inconvenient to explicitly shut down the server but for the processing-server-use-case it is inefficient

Solution 3) is as robust as the queue/cache are robust, so as long as there is no danger of stale jobs lingering, it might be the way to go.

Reply:
Forgot "solution 4" which is to just not handle shutdown of the METS server at all and hope that everything is shut down gracefully when the processing server is stopped.

Reply:
I have implemented Solution 3). It seems to work just fine.